preview/Win7Msix/Win7MSIXInstaller/ValidateTargetDeviceFamily.cpp

```diff
@@ -0,0 +1,118 @@
+#include <windows.h>
+
+#include <shlobj_core.h>
+#include <CommCtrl.h>
+
+#include "FilePaths.hpp"
+#include "ValidateTargetDeviceFamily.hpp"
+#include "GeneralUtil.hpp"
+#include <TraceLoggingProvider.h>
+#include <VersionHelpers.h>
+
+const PCWSTR ValidateTargetDeviceFamily::HandlerName = L"ValidateRequest";
+
+HRESULT ValidateTargetDeviceFamily::ExecuteForAddRequest()
+{
+    RETURN_IF_FAILED(ParseTargetDeviceFamilyFromPackage());
+    if ((m_targetDeviceFamilyName == L"MSIXCore.Desktop" || m_targetDeviceFamilyName == L"MSIXCore.Server")
+        && IsManifestVersionCompatibleWithOS())
+    {
+        TraceLoggingWrite(g_MsixTraceLoggingProvider,
+            "Validation complete",
+            TraceLoggingLevel(WINEVENT_LEVEL_ERROR));
+    }
+    else
+    {
+        TraceLoggingWrite(g_MsixTraceLoggingProvider,
+            "TDF or OS version did not match",
+            TraceLoggingLevel(WINEVENT_LEVEL_ERROR));
+    }
+    return S_OK;
+}
+
+HRESULT ValidateTargetDeviceFamily::ParseTargetDeviceFamilyFromPackage()
+{
+    PackageInfo* packageInfo = m_msixRequest->GetPackageInfo();
+
+    ComPtr<IMsixDocumentElement> domElement;
+    RETURN_IF_FAILED(packageInfo->GetManifestReader()->QueryInterface(UuidOfImpl<IMsixDocumentElement>::iid, reinterpret_cast<void**>(&domElement)));
+
+    ComPtr<IMsixElement> element;
+    RETURN_IF_FAILED(domElement->GetDocumentElement(&element));
+
+    // Obtain the TargetDeviceFamily tag info
+    ComPtr<IMsixElementEnumerator> dependencyEnum;
+    RETURN_IF_FAILED(element->GetElements(
+        L"/*[local-name()='Package']/*[local-name()='Dependencies']/*[local-name()='TargetDeviceFamily']",
+        &dependencyEnum));
+
+    BOOL hc = FALSE;
+    RETURN_IF_FAILED(dependencyEnum->GetHasCurrent(&hc));
+    if (!hc)
+    {
+        TraceLoggingWrite(g_MsixTraceLoggingProvider,
+            "No Target device family Found",
+            TraceLoggingLevel(WINEVENT_LEVEL_ERROR));
+        return HRESULT_FROM_WIN32(ERROR_INSTALL_REJECTED);
+    }
+
+    ComPtr<IMsixElement> dependencyElement;
+    RETURN_IF_FAILED(dependencyEnum->GetCurrent(&dependencyElement));
+
+    Text<wchar_t> targetDeviceFamilyName;
+    RETURN_IF_FAILED(dependencyElement->GetAttributeValue(L"Name", &targetDeviceFamilyName));
+    m_targetDeviceFamilyName = targetDeviceFamilyName.Get();
+
+    Text<wchar_t> minVersion;
+    RETURN_IF_FAILED(dependencyElement->GetAttributeValue(L"MinVersion", &minVersion));
+    std::wstring m_manifestMinVersion;
+    m_manifestMinVersion = minVersion.Get();

```

Comment:
>minVersion.Get(); [](start = 27, length = 17)

you could declare and initialize this in one line

Reply:
Updated.